host/sgx/enter.c

```diff
@@ -156,14 +159,21 @@ void oe_enter(
     uint64_t* arg4,
     oe_enclave_t* enclave)
 {
-    // The general purpose registers are preserved by the compiler.
-    // The floating point state and the flags must be explicitly preserved.
-    // The space for saving the floating-point state must be 16 byte aligned.
-    OE_ALIGNED(16)
-    uint64_t fx_state[64];
+    // Additional control registers that need to be preserved as part of the
+    // Windows and Linux x64 ABIs
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+
     oe_ecall_context_t ecall_context = {{0}};
     _setup_ecall_context(&ecall_context);
 
+    // Compiler will usually handle this on exiting a function that uses AVX,
+    // but we need to avoid the AVX-SSE transition penalty here manually as
+    // part of the transition to enclave. See https://software.intel.com/content
+    // /www/us/en/develop/articles/avoiding-avx-sse-transition-penalties.html
+    if (oe_is_avx_enabled)
```

Comment:
Do we enforce that when oe_enter is called, oe_is_avx_enabled would have been correctly initialized? I see that we initialize it using oe_once. But I'm curious about the case where enclaves and called via global initializers.

Reply:
I'm not sure I understand the scenario you have in mind using global initializers. `oe_is_avx_enabled` is globally initialized for the host process once on first invocation of any `oe_create_enclave` call, which is a prerequisite for calling an enclave method and `oe_enter`. Prior to this changeset, `oe_create_enclave` already had a dependency on `oe_once`, and this just reuses that mechanism for that additional initialization of `oe_is_avx_enabled`, so any interactions with the timing of global initialization already exists.

Reply:
Given that `oe_is_avx_enabled` is initialized as part of `oe_create_enclave` and not via a global constructor in which case ordering of global constructors would matter, the scenario I was curios about is addressed.